src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1317 @@
+                }
+
+                SolutionCompilationState? currentPartialSolution = null;
+                _latestSolutionWithPartialCompilation?.TryGetTarget(out currentPartialSolution);
+
+                var reuseExistingPartialSolution =
+                    (DateTime.UtcNow - _timeOfLatestSolutionWithPartialCompilation).TotalSeconds < 0.1 &&
+                    _documentIdOfLatestSolutionWithPartialCompilation == documentId;
+
+                if (reuseExistingPartialSolution && currentPartialSolution != null)
+                {
+                    SolutionLogger.UseExistingPartialSolution();
+                    return currentPartialSolution;
+                }
+
+                var newIdToProjectStateMap = this.Solution.ProjectStates;
+                var newIdToTrackerMap = _projectIdToTrackerMap;
+
+                foreach (var (doc, tree) in builder)
+                {
+                    // if we don't have one or it is stale, create a new partial solution
+                    var tracker = this.GetCompilationTracker(doc.Id.ProjectId);
+                    var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+
+                    Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(doc.Id.ProjectId));
+                    newIdToProjectStateMap = newIdToProjectStateMap.SetItem(doc.Id.ProjectId, newTracker.ProjectState);
+                    newIdToTrackerMap = newIdToTrackerMap.SetItem(doc.Id.ProjectId, newTracker);
+                }
+
+                var newState = this.Solution.Branch(
+                    idToProjectStateMap: newIdToProjectStateMap,
+                    dependencyGraph: SolutionState.CreateDependencyGraph(this.Solution.ProjectIds, newIdToProjectStateMap));
+                var newCompilationState = this.Branch(
+                    newState,
+                    newIdToTrackerMap);
+
+                _latestSolutionWithPartialCompilation = new WeakReference<SolutionCompilationState>(newCompilationState);
+                _timeOfLatestSolutionWithPartialCompilation = DateTime.UtcNow;
+                _documentIdOfLatestSolutionWithPartialCompilation = documentId;
+
+                SolutionLogger.CreatePartialSolution();
+                return newCompilationState;
+            }
+        }
+        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
+        {
+            throw ExceptionUtilities.Unreachable();
+        }
+    }
+
+    public SolutionCompilationState AddDocuments(ImmutableArray<DocumentInfo> documentInfos)
+    {
+        return AddDocumentsToMultipleProjects(documentInfos,
+            (documentInfo, project) => project.CreateDocument(documentInfo, project.ParseOptions, new LoadTextOptions(project.ChecksumAlgorithm)),
+            (oldProject, documents) => (oldProject.AddDocuments(documents), new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction(documents)));
```

Comment:
For follow up: I think all of these lambdas can be made static...and if they can't be that's kinda suspicious.

Reply:
done in followup.